File: examples/csharp/mega_flow/Storage.cs
namespace Mega;

// Class hierarchy — abstract base, inheritance, virtual + override,
// auto-properties, static factory — all preserving taint on the way
// to the sink.
public abstract class BaseRepository
{
    public Envelope Data { get; }

    protected BaseRepository(Envelope data)
    {
        Data = data;
    }

    // Expression-bodied property exposes the tainted cmd field.
    public string Cmd => Data.Cmd;

    public abstract string Run();
}

public class Repository : BaseRepository
{
    public Repository(Envelope data) : base(data) { }

    public static Repository Wrap(Envelope data) => new Repository(data);

    public override string Run()
    {
        var c = Cmd;
        return Executor.Execute(c);
    }
}

public class AuditedRepository : Repository
{
    public AuditedRepository(Envelope data) : base(data) { }

    public override string Run()
    {
        // base-call preserves taint across the inheritance chain.
        return base.Run();
    }
}

public static class Storage
{
    public static string Persist(Envelope envelope)
    {
        var repo = new AuditedRepository(envelope);
        return repo.Run();
    }
}

File: examples/kotlin/mega_flow/Storage.kt
package mega

typealias RepoEnvelope = App.Envelope

// Class hierarchy — open/abstract/override, properties with custom
// getters, sealed classes, companion-object factory — all preserving
// taint on the way to the sink.
sealed class RepoState {
    data object Active : RepoState()
}

abstract class BaseRepository(val data: RepoEnvelope) {
    private val state: RepoState = RepoState.Active

    // Custom getter exposes the tainted cmd field.
    open val cmd: String get() = data.cmd

    abstract fun run(): String
}

open class Repository(data: RepoEnvelope) : BaseRepository(data) {
    companion object {
        fun wrap(data: RepoEnvelope): Repository = Repository(data)
    }

    override fun run(): String {
        val c = cmd
        return Executor.execute(c)
    }
}

class AuditedRepository(data: RepoEnvelope) : Repository(data) {
    override fun run(): String {
        // super-call preserves taint across the inheritance chain.
        return super.run()
    }
}

object Storage {
    fun persist(envelope: RepoEnvelope): String {
        val repo = AuditedRepository(envelope)
        return repo.run()
    }
}
